src/testing/slim_check/sampleable.lean

```diff
@@ -0,0 +1,268 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Simon Hudon
+-/
+import data.lazy_list
+import data.lazy_list2
+import data.tree
+import tactic.linarith
+import testing.slim_check.gen
+
+/-!
+# Sampleable Class
+
+This class permits the creation samples of a given type
+controlling the size of those values using the `gen` monad`. It also
+helps minimize examples by creating smaller versions of given values.
+
+When testing a proposition like `∀ n : ℕ, prime n → n ≤ 100`,
+`slim_check` requires that `ℕ` have an instance of `sampleable` and for
+`prime n` to be decidable.  `slim_check` will then use the instance of
+`sampleable` to generate small examples of ℕ and progressively increase
+in size. For each example `n`, `prime n` is tested. If it is false,
+the example will be rejected (not a test success nor a failure) and
+`slim_check` will move on to other examples. If `prime n` is true, `n
+≤ 100` will be tested. If it is false, `n` is a counter-example of `∀
+n : ℕ, prime n → n ≤ 100` and the test fails. If `n ≤ 100` is true,
+the test passes and `slim_check` moves on to trying more examples.
+
+
+This is a port of the Haskell QuickCheck library.
+
+## Main definitions
+  * `sampleable` class
+
+## Tags
+
+random testing
+
+## References
+
+  * https://hackage.haskell.org/package/QuickCheck
+
+-/
+universes u
+
+namespace slim_check
+
+variables (α : Type u)
+
+/-- `sampleable α` provides ways of creating examples of type `α`,
+and given such an example `x : α`, gives us a way to shrink it
+and find simpler examples.  -/
+class sampleable :=
+(sample [] : gen α)
+(shrink : α → lazy_list α)
+
+export sampleable (sample shrink)
+
+open nat lazy_list
+
+/-- apply `f` to combine every element of the first list with every element
+of the second list and interleave the resulting lists
+
+For instance `lseq prod.mk [1,2,3] [4,5,6]` results in
+
+```
+[(1, 4), (2, 4), (1, 5), (3, 4), (1, 6), (2, 5), (3, 5), (2, 6), (3, 6)]
+```
+
+The purpose is to that two lists of shrunken values in ascending order of size
+and produce a list of combined values in roughly ascending order of size too.
+
+If we add the samples instead with `lseq (+) [1,2,3] [1,2,3]`, we
+obtain:
+
+```
+[2, 3, 3, 4, 4, 4, 5, 5, 6]
+```
+ -/
+def sampleable.lseq {α β γ} (f : α → β → γ) : lazy_list α → lazy_list β → lazy_list γ
+| lazy_list.nil xs := lazy_list.nil
+| (lazy_list.cons x xs) lazy_list.nil := lazy_list.nil
+| (lazy_list.cons x xs) ys := interleave (ys.map $ f x) (sampleable.lseq (xs ()) ys)
+
+/-- implementation of `sampleable nat` -/
```

Comment:
```suggestion
/--
[describe here what it does]

This function is used to implement `sampleable.nat.shrink`.
-/
```

Reply:
The description you refer to is in the doc string of `nat.shrink`

Reply:
Yes, please duplicate it.

Reply:
I can duplicate it but it seems like a bad idea to me, I think that risks introducing inconsistent documentation.

Reply:
You can cross reference the docs if you want to write a more elaborate version, but you should have at least a basic documentation at all locations, even if it is only copy pasta. I experienced this with `rcases`; the possible rcases patterns are described with varying levels of detail in at least three places there.